predict the closest cluster belonging to input vector

Hello,
I used kmeans clustering(elkan) algorithm with mahalanobis distance to create around 2000 cluster for my application. It took me alot of time to get these cluster, its centroids. 
I want to know cluster assignment for new data values but there is no function available wherein there would be no training and just cluster assignment.
Please help me how shall I do this with available functionality???

Hello,
Thanks for the suggestion.
I have tried the above suggestion and used k nearest neighbor. But
unfortunately it didn't worked with Mahalanobis Distance metric type.
Mahalanobis distance dont work with tree algorithms, I have tried with
kmeans and dualtree algorithm. With both I have got this error!!!

/usr/local/include/mlpack/core/tree/hrectbound_impl.hpp:243:27: error:
‘Power’ is not a member of ‘mlpack::metric::MahalanobisDistance<true>’
     if (MetricType::Power == 1)
                           ^
/usr/local/include/mlpack/core/tree/hrectbound_impl.hpp:245:32: error:
‘Power’ is not a member of ‘mlpack::metric::MahalanobisDistance<true>’
     else if (MetricType::Power == 2)
                                ^
/usr/local/include/mlpack/core/tree/hrectbound_impl.hpp:252:17: error:
‘Power’ is not a member of ‘mlpack::metric::MahalanobisDistance<true>’
       sum += pow((lower + fabs(lower)) + (higher + fabs(higher)),
                 ^
/usr/local/include/mlpack/core/tree/hrectbound_impl.hpp:262:25: error:
‘Power’ is not a member of ‘mlpack::metric::MahalanobisDistance<true>’
   if (MetricType::Power == 1)
                         ^
/usr/local/include/mlpack/core/tree/hrectbound_impl.hpp:264:30: error:
‘Power’ is not a member of ‘mlpack::metric::MahalanobisDistance<true>’
   else if (MetricType::Power == 2)
                              ^
/usr/local/include/mlpack/core/tree/hrectbound_impl.hpp:266:5: error:
‘TakeRoot’ is not a member of ‘mlpack::metric::MahalanobisDistance<true>’
     if (MetricType::TakeRoot)
     ^
/usr/local/include/mlpack/core/tree/hrectbound_impl.hpp:273:5: error:
‘TakeRoot’ is not a member of ‘mlpack::metric::MahalanobisDistance<true>’
     if (MetricType::TakeRoot)
     ^
/usr/local/include/mlpack/core/tree/hrectbound_impl.hpp:275:15: error:
‘Power’ is not a member of ‘mlpack::metric::MahalanobisDistance<true>’
           1.0 / (double) MetricType::Power) / 2.0;
               ^
/usr/local/include/mlpack/core/tree/hrectbound_impl.hpp:277:23: error:
‘Power’ is not a member of ‘mlpack::metric::MahalanobisDistance<true>’
       return sum / pow(2.0, MetricType::Power);


On Thu, Jun 28, 2018 at 7:10 PM, Ryan Curtin <notifications@github.com>
wrote:

> Hey there,
>
> If Elkan's algorithm was slow I might recommend trying Hamerly's algorithm
> or one of the tree-based algorithm. The dualtree algorithm may be useful if
> the data is of low(ish) dimensionality, since you are searching for many
> clusters.
>
> My suggestion would be to use k-nearest-neighbor search for this, with the
> cluster centroids as your reference set and your new points as the query
> set. That will return to you the indices of the nearest cluster centroid,
> which is exactly the cluster assignment of each point that you are looking
> for.
>
> I hope this helps! Let me know if I can clarify.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/mlpack/mlpack/issues/1453#issuecomment-401038266>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AefwXubSOf4mK-vFipxHnhaJ_1OxWVLDks5uBNzOgaJpZM4U7Xlm>
> .
>

--- a/.travis.yml
+++ b/.travis.yml
@@ -13,8 +13,8 @@ before_install:
   # For the python bindings we need pandas >= 0.15.0.
   - wget -O- http://neuro.debian.net/lists/trusty.us-ca.full | sudo tee /etc/apt/sources.list.d/neurodebian.sources.list
   - sudo apt-key adv --recv-keys --keyserver hkp://ha.pool.sks-keyservers.net 0xA5D32F012649A5A9
-  - sudo apt-get update -qq
-  - sudo apt-get install -qq libopenblas-dev liblapack-dev g++ libboost-all-dev python3-pip cython3 python3-numpy python3-pandas
+  - sudo apt-get update
+  - sudo apt-get install -y --allow-unauthenticated libopenblas-dev liblapack-dev g++ libboost-all-dev python3-pip cython3 python3-numpy python3-pandas
   # Install both python2 and python3 modules, and the build will decide which to
   # use.
   - sudo pip install cython numpy pandas

--- a/src/mlpack/core/metrics/mahalanobis_distance.hpp
+++ b/src/mlpack/core/metrics/mahalanobis_distance.hpp
@@ -31,9 +31,16 @@ namespace metric {
  * where Q is the covariance matrix.
  *
  * Because each evaluation multiplies (x_1 - x_2) by the covariance matrix, it
- * may be much quicker to use an LMetric and simply stretch the actual dataset
- * itself before performing any evaluations.  However, this class is provided
- * for convenience.
+ * is typically much quicker to use an LMetric and simply stretch the actual
+ * dataset itself before performing any evaluations.  However, this class is
+ * provided for convenience.
+ *
+ * If you wish to use the KNN class or other tree-based algorithms with this
+ * distance, it is recommended to instead stretch the dataset first, by
+ * decomposing Q = L^T L (perhaps via a Cholesky decomposition), and then
+ * multiply the data by L.  If you still wish to use the KNN class with a custom
+ * distance anyway, you will need to use a different tree type than the default
+ * KDTree, which only works with the LMetric class.
  *
  * Similar to the LMetric class, this offers a template parameter TakeRoot
  * which, when set to false, will instead evaluate the distance

